### PR TITLE
✨ 9.03/9.04 Support

### PR DESCRIPTION
**Pull request by iMrDJAi**
Added offsets for FW 9.03. Gadgets are mostly found by @zecoxao.

I dedicate this one to our friend [@903jailbreak](https://twitter.com/903jailbreak) 😂

image

### CODE REVIEW

**Comment by iMrDJAi** [comment]
I need testers btw

**Comment by dxcool222** [comment]
I’ll test here in a bit I have a PS4 Pro on 9.03

**Comment by AkoposiNobi** [comment]
I would love to test this out but im on 9.04 xD

**Comment by hydroxideHO21** [comment]
> I need testers btw

I have a fat 9.03 I can test now

**Comment by W-i-n-7** [comment]
[PPPwn903.zip](https://github.com/TheOfficialFloW/PPPwn/files/15178469/PPPwn903.zip)
precompiled by me

**Comment by W-i-n-7** [comment]
@DJTOMATO here

> I wanted to test it in 9.03, PS4 pro too Can anyone share the compiled files? I cannot compile on windows, neither on my rpi/arm

**Comment by UndeadTigerAU** [comment]
Am willing to test as well!

**Comment by W-i-n-7** [comment]
im on 10.50 :/

**Comment by AmineSimcos** [comment]
im on 11.50 :(

**Comment by MagmaSKV** [comment]
I have 10.00, if you need a tester for that one, I could try to help you.

**Comment by W-i-n-7** [comment]
> im on 11.50 :(

you can leave...
not to be rude

**Comment by DJTOMATO** [comment]
```[+] PPPwn - PlayStation 4 PPPoE RCE by theflow
[+] args: interface=eth0 fw=903 stage1=stage1/stage1.bin stage2=stage2/stage2.b         
in

[+] STAGE 0: Initialization
[*] Waiting for PADI...
[+] pppoe_softc: 0xffffec1b077b4800
[+] Target MAC: 2c:cc:44:3d:xx:xx <= Double checked and it's ps4 pro MAC, so it finds it properly
[+] Source MAC: 07:48:7b:07:xx:xx
[+] AC cookie length: 0x4e0
[*] Sending PADO...
[*] Waiting for PADR...
```

CE-33984-7 Cannot obtain an ip adderss within the time limit

**Comment by Clompress** [comment]
Maybe the next one will be 10.71? :)

**Comment by se2crid** [comment]
how are you getting the offsets? and how are you getting them if you dont have a 9.03 ps4?

**Comment by hydroxideHO21** [comment]
i got further
image

**Comment by lompaket** [comment]
I get to the same place using the 10.01 offsets

**Comment by hydroxideHO21** [comment]
UPDATE got it to here and froze my PS4
image
will continue brute forcing lol

**Comment by kaz2700** [comment]
Also got to the stage1
Screenshot_20240501_184735 and PS4 froze (9.03) firmware

**Comment by hydroxideHO21** [comment]
@iMrDJAi Seems to continually get stuck at "Waiting for stage1 to resume..."
I've tried probably 50 times and this is as far as it gets

Hope this helps!!

**Comment by hydroxideHO21** [comment]
what I've seen so far with my console (PS4 Fat)
- 7/10 times fails to find corrupted object
- 2/10 times gets stuck waiting for LCP configure object
- 1/10 times gets stuck waiting for stage 1 to resume after getting to stage 3 (causes ps4 to hard freeze, controllers to disconnect, and must be unplugged to reset)

**Comment by hydroxideHO21** [comment]
from what I can determine, seems to be an offsets issue. wish I knew how to dump my kernel

**Comment by se2crid** [comment]
> The only thing that I haven't double check is the gadgets. I'm going to search for them again using https://github.com/JonathanSalwan/ROPgadget.

what are the gadgets?

**Comment by JoElH4ck3r2022** [comment]
offsets 10.00??

**Comment by se2crid** [comment]
> offsets 10.00??

nobody is working on them rn but that fw can be jailbroken

**Comment by zecoxao** [comment]
just a note here that the gadgets from 9.03 are identical to the ones on 9.04

**Comment by Bryan-De** [comment]
Doesn't work my console this freeze

**Comment by Cracko298** [comment]
Just tested, works great

**Comment by amraj007** [comment]
Please do the 10.70

**Comment by HiDeath98** [comment]
> Just tested, works great

what PS4 version do you have exactly,
can you please give the exact steps you did?

**Comment by se2crid** [comment]
> Just tested, works great

HOW?

**Comment by hydroxideHO21** [comment]
I have B03 FAT PS4

**Comment by ablixM** [comment]
> i got further @iMrDJAi image sometimes i get stuck here image

same for me fat ps4 9.03

**Comment by se2crid** [comment]
> I have B03 FAT PS4

you mean 9.03 right?

**Comment by hydroxideHO21** [comment]
> > I have B03 FAT PS4
> 
> 
> 
> you mean 9.03 right?
> 
> 

Yes 😂

**Comment by se2crid** [comment]
what happens when you run the script

**Comment by hydroxideHO21** [comment]
> what happens when you run the script



> what I've seen so far with my console (PS4 Fat)
> 
> * 7/10 times fails to find corrupted object
> * 2/10 times gets stuck waiting for LCP configure object
> * 1/10 times gets stuck waiting for stage 1 to resume after getting to stage 3 (causes ps4 to hard freeze, controllers to disconnect, and must be unplugged to reset)

**Comment by UndeadTigerAU** [comment]
Hey I want to test it out on my PS4 but I'm unsure how to, any help appreciated.

**Comment by Jara1596** [comment]
> Also got to the stage1 a couple of times and PS4 froze (9.03 pro) firmware Screenshot_20240501_184735

Same for me using ps4 slim (9.03)

**Comment by iMrDJAi** [comment]
Need testers ^^^^^

**Comment by Bryan-De** [comment]
Can you send me the decompile files I can't decompile them stage 1 and 2

**Comment by Bryan-De** [comment]
> Need testers ^^^^^

[+] STAGE 0: Initialization
[*] Waiting for PADI...
[+] pppoe_softc: 0xffffb72d07e84a00
[+] Target MAC: 0c:fe:45:45:97:c1
[+] Source MAC: 07:4a:e8:07:2d:b7
[+] AC cookie length: 0x4e0
[*] Sending PADO...
[*] Waiting for PADR...
[*] Sending PADS...
[*] Sending LCP configure request...
[*] Waiting for LCP configure ACK...
[*] Waiting for LCP configure request...
[*] Sending LCP configure ACK...
[*] Sending IPCP configure request...
[*] Waiting for IPCP configure ACK...
[*] Waiting for IPCP configure request...
[*] Sending IPCP configure NAK...
[*] Waiting for IPCP configure request...
[*] Sending IPCP configure ACK...
[*] Waiting for interface to be ready...
[+] Target IPv6: fe80::efe:45ff:fe45:97c1
[+] Heap grooming...done

[+] STAGE 1: Memory corruption
[+] Pinning to CPU 0...done
[*] Sending malicious LCP configure request...
[*] Waiting for LCP configure reject...
[*] Sending LCP configure request...
[*] Waiting for LCP configure ACK...
[*] Waiting for LCP configure request...
[*] Sending LCP configure ACK...
[*] Sending IPCP configure request...
[*] Waiting for IPCP configure ACK...
[*] Waiting for IPCP configure request...
[*] Sending IPCP configure NAK...
[*] Waiting for IPCP configure request...
[*] Sending IPCP configure ACK...
[+] Scanning for corrupted object...found fe80::092f:4141:4141:4141

[+] STAGE 2: KASLR defeat
[*] Defeating KASLR...
[+] pppoe_softc_list: 0xffffffffcf7f19f8
[+] kaslr_offset: 0x4b408000

[+] STAGE 3: Remote code execution
[*] Sending LCP terminate request...
[*] Waiting for PADI...
[+] pppoe_softc: 0xffffb72d07e84a00
[+] Target MAC: 0c:fe:45:45:97:c1
[+] Source MAC: 6f:e8:de:cd:ff:ff
[+] AC cookie length: 0x514
[*] Sending PADO...
[*] Waiting for PADR...
[*] Sending PADS...
[*] Triggering code execution...
[*] Waiting for stage1 to resume...

does not work for me it is completely buggy the console menus impossible to turn it off in part by unplugging the console

**Comment by kaz2700** [comment]
Can confirm this behaviour 
Screenshot_20240502_014025

**Comment by Jara1596** [comment]
I've got the same bug, console menu is buggy, no sound and after a while it crash.

**Comment by kaz2700** [comment]
Maybe this time its the stage 1 offsets :thinking:

**Comment by dxcool222** [comment]
Same here sound goes away then system crashes on my PS4 Pro

**Comment by iMrDJAi** [comment]
damn it! sorry! my bad..

**Comment by iMrDJAi** [comment]
POP_R8_POP_RBP_RET is cursed! Anyway...
People on 9.03/9.04 please test again. Sorry again.

**Comment by Bryan-De** [comment]
> POP_R8_POP_RBP_RET is cursed! Anyway...
> 
> People on 9.03/9.04 please test again. Sorry again.

Does not work when the console is completely turned off

**Comment by iMrDJAi** [comment]
@Bryan-De We got a new imposter in this case. Getting back to work...

**Comment by dxcool222** [comment]
First time it failed 2nd time system shut off, I’ll keep testing

**Comment by Bryan-De** [comment]
> @Bryan-De We got a new imposter in this case. Getting back to work...

should we decompile stage 1 and 2 with each new update?

**Comment by DJTOMATO** [comment]
can anyone precompile for 9.03?

**Comment by dxcool222** [comment]
I’ll be waiting :)

**Comment by dxcool222** [comment]
> > @Bryan-De We got a new imposter in this case. Getting back to work...
> 
> should we decompile stage 1 and 2 with each new update?

Every time a fix Is dropped I delete my old files and redo everything just to be safe, it take like 5 sec

**Comment by HiDeath98** [comment]
Willing to test this,
I just need the steps apart from the ones listed in here: https://github.com/TheOfficialFloW/PPPwn

**Comment by Bryan-De** [comment]
> > > @Bryan-De We got a new imposter in this case. Getting back to work...
> > 
> > 
> > should we decompile stage 1 and 2 with each new update?
> 
> Every time a fix Is dropped I delete my old files and redo everything just to be safe, it take like 5 sec

windows or linux ???

File: offsets.py [hunk 1]
```diff
@@ -150,7 +151,7 @@ class OffsetsFirmware_903:
     POP_RCX_RET = 0xFFFFFFFF822445E7 # 59 C3
 
     # pop r8 ; pop rbp ; ret
-    POP_R8_POP_RBP_RET = 0xFFFFFFFF822AB4DD # 47 58 5D C3
+    POP_R8_POP_RBP_RET = 0xffffffff8237dd7d # C5 FB 11 47 58 5D C3
```
Comment (loskutov) [inline offsets.py]: In fact you don't want the "C5 FB 11" part, so the comment is kind of misleading..

**Comment by dxcool222** [comment]
> > > > @Bryan-De We got a new imposter in this case. Getting back to work...
> > > 
> > > 
> > > should we decompile stage 1 and 2 with each new update?
> > 
> > 
> > Every time a fix Is dropped I delete my old files and redo everything just to be safe, it take like 5 sec
> 
> windows or linux ???

I am using Linux mint

**Comment by Jara1596** [comment]
ps4 slim 9.03 
system shutoff, tried many times

**Comment by Bryan-De** [comment]
> > > > > @Bryan-De We got a new imposter in this case. Getting back to work...
> > > > 
> > > > 
> > > > should we decompile stage 1 and 2 with each new update?
> > > 
> > > 
> > > Every time a fix Is dropped I delete my old files and redo everything just to be safe, it take like 5 sec
> > 
> > 
> > windows or linux ???
> 
> I am using Linux mint

I have an error

image

**Comment by Jara1596** [comment]
> > > > > > @Bryan-De We got a new imposter in this case. Getting back to work...
> > > > > 
> > > > > 
> > > > > should we decompile stage 1 and 2 with each new update?
> > > > 
> > > > 
> > > > Every time a fix Is dropped I delete my old files and redo everything just to be safe, it take like 5 sec
> > > 
> > > 
> > > windows or linux ???
> > 
> > 
> > I am using Linux mint
> 
> I have an error
> 
> image

seems ur freebds-headers are empty, [freebsd-headers](https://github.com/OpenOrbis/freebsd-headers) , u need them and it will work

**Comment by dxcool222** [comment]
> > > > > > @Bryan-De We got a new imposter in this case. Getting back to work...
> > > > > 
> > > > > 
> > > > > should we decompile stage 1 and 2 with each new update?
> > > > 
> > > > 
> > > > Every time a fix Is dropped I delete my old files and redo everything just to be safe, it take like 5 sec
> > > 
> > > 
> > > windows or linux ???
> > 
> > 
> > I am using Linux mint
> 
> I have an error
> 
> image

Make sure to follow the guide, I had this error and I just used the git command instead of downloading it and it fixed it.. git clone --recursive https://github.com/iMrDJAi/PPPwn.git

**Comment by dxcool222** [comment]
Also I have to type FW900 if I type FW903 it will say invalid firmware

**Comment by hydroxideHO21** [comment]
> ps4 slim 9.03 
> 
> system shutoff, tried many times

Do a git pull and then do a gh Cl

**Comment by hydroxideHO21** [comment]
> ps4 slim 9.03 
> 
> system shutoff, tried many times

Wrong reply lol

**Comment by hydroxideHO21** [comment]
> Willing to test this,
> 
> I just need the steps apart from the ones listed in here: https://github.com/TheOfficialFloW/PPPwn

Do git pull then gh cl this authors changes

File: offsets.py (hunk 1 above)
Comment (iMrDJAi) [reply offsets.py]: @zecoxao

Comment (iMrDJAi) [reply offsets.py]: @loskutov I guess I should go to bed at this point. Man, how did I do that? 🤦‍♂️

**Comment by iMrDJAi** [comment]
image

**Comment by dxcool222** [comment]
Damn my system still crashing

**Comment by iMrDJAi** [comment]
> Also I have to type FW900 if I type FW903 it will say invalid firmware

@dxcool222 What? this is wrong, you gotta type `--fw=903`.

**Comment by DJTOMATO** [comment]
Can anyone build the files please? I can't on my RPI (ARM64) and windows just won't work

**Comment by Jara1596** [comment]
> image

ps4 slim (9.03) i got PPPwned too
Thanks and GN

**Comment by HiDeath98** [comment]
Someone please upload the final compiled version,
As well as some steps on how to get to work

**Comment by DJTOMATO** [comment]
I've recompiled the files again but i still get stuck here 

```
[+] Target MAC: 2c:cc:44:3d:xx
[+] Source MAC: 07:84:b2:28:xx
[+] AC cookie length: 0x4e0      
[*] Sending PADO...
[*] Waiting for PADR...

```

**Comment by HiDeath98** [comment]
I'm getting this:
image

While trying to run "make -C stage1 FW=903 clean && make -C stage1 FW=903"

**Comment by dxcool222** [comment]
> I'm getting this: image
> 
> While trying to run "make -C stage1 FW=903 clean && make -C stage1 FW=903"

Me too

**Comment by HiDeath98** [comment]
> > I'm getting this: image
> > While trying to run "make -C stage1 FW=903 clean && make -C stage1 FW=903"
> 
> Me too

You have to install the original from here: https://github.com/TheOfficialFloW/PPPwn
Right after installing, install gh (apt-get),
Then run these:
cd PPPwn
gh auth login

then follow the steps to login into your github account, you can use the browser, make sure to copy the login code before.
then run the following:
gh pr checkout 15

it should work normally after this, make sure to replace 1100 with 903 for all the remaining steps.

**Comment by DJTOMATO** [comment]
hcSjEKALMUGg
I am using a different USB adapter and now I get stuck here, no pop-up/notification.
What could it be?

**Comment by thecodefodder** [comment]
Has anyone encountered their console freezing when the exploit reaches "Waiting for stage1 to resume..."?

(PS4 Fat)

**Comment by DJTOMATO** [comment]
Code_qgGi4s79dU

- Confirmed working on ps4 pro, 9.03
- Works best when made on test internet connection AFTER setting up
- "Ethernet 2" won't work as parameter when using a USB adapter, it's best to disable all other interfaces
-

**Comment by Kohese** [comment]
Got it working on 9.04 after about 10 attempts

**Comment by dxcool222** [comment]
> Got it working on 9.04 after about 10 attempts

Man that’s a lot of tries I wonder what the best stable firmware for this exploit, guess we will find out

**Comment by RotaryGabry** [comment]
It works on my 9.03 PS4 Pro. For me it worked 3 out of 4 attempts, i once get an error scanning for corrupted object.

File: offsets.py [hunk 2]
```diff
@@ -98,6 +99,100 @@ class OffsetsFirmware_900:
...
+    SETIDT = 0xffffffff825128e0
+
+    KMEM_ALLOC = 0xffffffff8257a070
+    KMEM_ALLOC_PATCH1 = 0xffffffff8257a13c
+    KMEM_ALLOC_PATCH2 = 0xffffffff8257a144
+
+    MEMCPY = 0xffffffff82471130
+
+    # mov cr0 rsi ; ud2 ; mov eax 1; ret
+    MOV_CR0_RSI_UD2_MOV_EAX_1_RET = 0xFFFFFFFF823FB679 # 0F 22 C6 48 F7 C6
```
Comment (TheOfficialFloW) [inline offsets.py:116]: Use lower case for hex and remove the bytes.

File: offsets.py [hunk 3]
```diff
@@ -98,6 +99,100 @@ class OffsetsFirmware_900:
...
+
+    SECOND_GADGET_OFF = 0x3D
+
+    # jmp qword ptr [rsi + 0x3d]
```
Comment (TheOfficialFloW) [inline offsets.py:120]: Use ROPgadget, put gadget in comment and use the address in lower case.

File: stage2/offsets.h [hunk 4]
```diff
@@ -53,6 +76,28 @@
 #define kdlsym_addr_copyinstr_patch2 0xffffffff824de4ef
 #define kdlsym_addr_copyinstr_patch3 0xffffffff824de520
 
+#elif FIRMWARE == 1001 // FW 10.01
```
Comment (TheOfficialFloW) [inline stage2/offsets.h:79]: Remove this.

**Comment by Bryan-De** [comment]
work for me thanks for the job !

File: offsets.py (hunk 3 above)
Comment (iMrDJAi) [reply offsets.py:120]: > put gadget in comment

Instruction unclear, what do you mean here?

File: offsets.py [hunk 5]
```diff
@@ -99,7 +99,7 @@ class OffsetsFirmware_900:
     JMP_R14 = 0xffffffff82b85693
 
 
-# FW 9.03
+# FW 9.03/9.04
 class OffsetsFirmware_903:
```
Comment (TheOfficialFloW) [inline offsets.py:103]: Call this OffsetsFirmware_903_94

File: offsets.py (hunk 3 above)
Comment (TheOfficialFloW) [reply offsets.py:120]: Like 0xffffffff82996603 : jmp qword ptr [rsi + 0x3d] instead of jmp qword ptr [rsi + 0x3d]

File: offsets.py (hunk 5 above)
Comment (theblackethiopiandude) [reply offsets.py:103]: OffsetsFirmware_903_94 or OffsetsFirmware_903_04?

**Comment by noamb14** [comment]
This version works with windows too or only linux?

**Comment by iMrDJAi** [comment]
@TheOfficialFloW Forgot to update README 💀

**Comment by se2crid** [comment]
> @TheOfficialFloW Forgot to update README 💀

yeah

**Comment by fffoo** [comment]
Yeah

**Comment by iMrDJAi** [comment]
@kaz2700 are you on latest?

**Comment by AkoposiNobi** [comment]
I've tested this to my fw9.04 PS4 and I can't make it display PPPwned.

Sometimes fails to find corrupted object,
sometimes gets stuck waiting for LCP configure object, and once in a while getting to stage 4 and crash the console when the stage2.bin is successfully injected.

**Comment by zecoxao** [comment]
it's normal that it fails to find the corrupted object on windows. it happens to me a lot as well, sometimes it finds it after 3 tries, sometimes after one try, sometimes after 10 or 20 tries.

**Comment by AkoposiNobi** [comment]
> it's normal that it fails to find the corrupted object on windows. it happens to me a lot as well, sometimes it finds it after 3 tries, sometimes after one try, sometimes after 10 or 20 tries.

But once it succeeded and proceed to stage 4 and sends payload to my PS4, my PS4 turns off

**Comment by zecoxao** [comment]
> > it's normal that it fails to find the corrupted object on windows. it happens to me a lot as well, sometimes it finds it after 3 tries, sometimes after one try, sometimes after 10 or 20 tries.
> 
> But once it succeeded and proceed to stage 4 and sends payload to my PS4, my PS4 turns off

compile stage1 and stage2 properly and you will have no issues

**Pull request merged**